### ISSUE TITLE
Returned headers differ per backend

### ISSUE THREAD

**Issue by Octogonapus** [JuliaCloud/AWS.jl#547]
When using `"return_headers" => true` as a parameter, the returned headers depend on the backend implementation. For example, the HTTP backend returns:
```
  "Date"                => "Thu, 05 May 2022 18:31:02 GMT"
  "x-amz-bucket-region" => "us-east-2"
  "Content-Type"        => "application/xml"
  "Transfer-Encoding"   => "chunked"
  "Server"              => "AmazonS3"
```

while the Downloads backend returns:
```
  "date"                => "Thu, 05 May 2022 18:30:44 GMT"
  "x-amz-bucket-region" => "us-east-2"
  "content-type"        => "application/xml"
  "transfer-encoding"   => "chunked"
  "server"              => "AmazonS3"
```

This means that any API consumers that need the header information must implement their code twice to accommodate each backend. As far as I can tell, the backend abstraction is otherwise not leaky, so if this header case issue could be addressed, then code could use this API generically without regard for the backend (also ignoring concurrency issues with HTTP).

**Comment by ericphanson** [comment]
I think this happens here in Downloads.jl: https://github.com/JuliaLang/Downloads.jl/blob/9f738d30e1256a4c122dff9f38536cfc1feeca8e/src/Curl/Easy.jl#L285

They lowercase the returned headers. Not sure why. We could also lowercase the HTTP.jl headers to match? But that's probably breaking

**Comment by ericphanson** [comment]
The commit message that introduced that code explains:

> The headers are returned with their keys lowercased. Doing this now is
future-proofing any users of this since HTTP/2 headers are always
lowercase and many users are probably incorrectly checking for case-
sensitive headers like "Content-Type" which are going to be returned
as "content-type" from an HTTP/2 server. We may as well get ahead of
the curve on this one since we have no users yet.

https://github.com/JuliaLang/Downloads.jl/commit/b1769519154605b0b52290aac256751e0fef621a

**Comment by ericphanson** [comment]
We could add an option in the HTTPBackend struct to lowercase the headers, and then later make it the default once we have a breaking release.

**Comment by mattBrzezinski** [comment]
> We could also lowercase the HTTP.jl headers to match? But that's probably breaking.

Definitely a breaking change to make. AWS themselves also expect certain casings for certain requests, such as [S3 multi-part uploading](https://github.com/JuliaCloud/AWSS3.jl/blob/067672b746aef7d4e37a3880e45ccbf20fce189b/src/AWSS3.jl#L853-L857). 

Maybe we can make a PR to `Downloads.jl` where you can pass in a kwarg to preserve the casings?

**Comment by ericphanson** [comment]
From the commit message I quoted above it sounds like this is deliberate and that we’ll run into HTTP2 issues if we rely on casing so I think the other solutions are better.

**Comment by mattBrzezinski** [comment]
> From the commit message I quoted above it sounds like this is deliberate and that we’ll run into HTTP2 issues if we rely on casing so I think the other solutions are better.

Yeah it seems deliberate, but the kwarg gives the option to do what you want, and gives backwards compat with HTTP/1 

FeatureSet idea works well and is least disruptive to others. Just seems annoying if depending on what you’re doing if mixing with AWSS3.

**Comment by ericphanson** [comment]
Well we could use that FeatureSet option in AWSS3’s `@service S3` call so it always sees lowercase headers. That might be breaking though, since it changes the return of `s3_get_meta`.

The other issue with a Downloads PR is it’s a stdlib (so it’s tied to Julia releases) and features aren’t backported so the soonest we could get it is Julia 1.9.

**Comment by mattBrzezinski** [comment]
> Well we could use that FeatureSet option in AWSS3’s `@service S3` call so it always sees lowercase headers. That might be breaking though, since it changes the return of `s3_get_meta`.

I'll spend some more time this week thinking about this and see what non-breaking change we can make here. If worst comes to worst we could do `AWSS3@2`, but would need to consider what else we want to break here with that.

I saw there was some discussion around this package on Slack, don't know if @omus @ExpandingMan @quinnj @iamed2 would want to change here with that.

> The other issue with a Downloads PR is it’s a stdlib (so it’s tied to Julia releases) and features aren’t backported so the soonest we could get it is Julia 1.9.

Ah, right. In my mind it was a separate package on its own.

**Comment by ericphanson** [comment]
AWSS3 isn’t 1.0 yet, by the way

**Comment by mattBrzezinski** [comment]
> AWSS3 isn’t 1.0 yet, by the way

We had an discussion at Invenia about moving everything to `1.0` when stable, guess this package slipped through some cracks. Guess it doesn't matter too much, could just introduce these breaking changes and finally get this moved to a major version.
